gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeCatalog.java

```diff
@@ -0,0 +1,393 @@
+    }
+
+    return true;
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
+    String fromDbId = databaseId(from.namespace());
+    String toDbId = databaseId(to.namespace());
+
+    Preconditions.checkArgument(
+        fromDbId.equals(toDbId),
+        "Cannot rename table %s to %s: database must match",
+        from.toString(),
+        to.toString());
+    bigLakeClient.renameTable(tableName(fromDbId, from.name()), tableName(toDbId, to.name()));
+  }
+
+  @Override
+  public void createNamespace(Namespace namespace, Map<String, String> metadata) {
+    if (namespace.isEmpty()) {
+      // Used by `CREATE NAMESPACE <catalog>`. Create a BLMS catalog linked with Iceberg catalog.
+      bigLakeClient.createCatalog(catalogName, Catalog.getDefaultInstance());
+      LOG.info("Created BigLake catalog: {}", catalogName.toString());
+    } else if (namespace.levels().length == 1) {
+      // Create a database.
+      String dbId = databaseId(namespace);
+      Database.Builder builder = Database.newBuilder().setType(Database.Type.HIVE);
+      builder
+          .getHiveOptionsBuilder()
+          .putAllParameters(metadata)
+          .setLocationUri(databaseLocation(dbId));
+
+      bigLakeClient.createDatabase(
+          DatabaseName.of(bigLakeProjectId, bigLakeRegion, catalogId, dbId), builder.build());
+    } else {
+      throw new IllegalArgumentException(
+          String.format("Invalid namespace (too long): %s", namespace));
+    }
+  }
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    if (!namespace.isEmpty()) {
+      // BLMS does not support namespaces under database or tables, returns empty.
+      // It is called when dropping a namespace to make sure it's empty, returns empty to unblock
+      // deletion.
+      return ImmutableList.of();
+    }
+
+    return Streams.stream(bigLakeClient.listDatabases(catalogName))
+        .map(BigLakeCatalog::namespace)
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  @Override
+  public boolean dropNamespace(Namespace namespace) {
+    try {
+      if (namespace.isEmpty()) {
+        // Used by `DROP NAMESPACE <catalog>`. Deletes the BLMS catalog linked by Iceberg catalog.
+        bigLakeClient.deleteCatalog(catalogName);
+        LOG.info("Deleted BigLake catalog: {}", catalogName.toString());
+      } else if (namespace.levels().length == 1) {
+        bigLakeClient.deleteDatabase(databaseName(namespace));
+        // Don't delete the data file folder for safety. It aligns with HMS's default behavior.
+        // To support database or catalog level config controlling file deletion in future.
+      } else {
+        LOG.warn("Invalid namespace (too long): {}", namespace);
+        return false;
+      }
+    } catch (NoSuchNamespaceException e) {
+      return false;
+    }
+
+    return true;
+  }
+
+  @Override
+  public boolean setProperties(Namespace namespace, Map<String, String> properties) {
```

Comment:
I think this should handle `location` separately because that's a property that's stored in the `Database` object.

Reply:
Removed the additional property "location".